### ISSUE TITLE
[RFC] Support Anticipatory Routing to mitigate MoE training loss spikes

### ISSUE THREAD

**Issue by chango9543** [pytorch/torchtitan#4349]
## Summary

This RFC proposes adding support for **Anticipatory Routing**, a training-stability technique
described in the DeepSeek-V4 technical [report,](https://arxiv.org/pdf/2606.19348) to torchtitan. The core idea is to
*decouple the synchronous updates of the backbone network and the routing network* during
MoE training, so that routing decisions are made from a slightly stale parameter state while
feature computation still uses the current state.

The motivation is to reduce the frequency/sharpness of **loss spikes** that arise during MoE
training, at a bounded and controllable additional training cost.

## Background & Motivation

MoE models route each token to a subset of experts through a routing (gating) network. When the
routing network and the backbone are updated synchronously, certain training regimes can become
unstable and produce sharp loss spikes. These spikes are disruptive: they often require a rollback
and can otherwise derail the run or require conservative hyper-parameter choices.

The DeepSeek-V4 technical report identifies this as a coupling problem and proposes
Anticipatory Routing as a remedy. This is a well-defined, non-trivial technique that is a good
candidate for torchtitan, which already contains the building blocks (MoE + Expert Parallelism,
pipeline parallelism, checkpointing) needed to express it.

## Proposed Technique (as described in the paper)

At training step `t`:

- **Feature computation** uses the current network parameters `θ_t`.
- **Routing indices** are computed and applied using historical parameters `θ_{t-Δt}`.

To avoid the cost of loading the model twice:

- The data for step `t` is fetched in advance **at step `t-Δt`**.
- The routing indices are **anticipated** (computed and cached) at step `t-Δt`, then replayed at
  step `t`. This is why it's called "Anticipatory Routing".

Infrastructure-level optimizations noted in the paper:

- Pre-computing routing indices requires **only a single forward pass** over the data (no
  backward pass), so it has low memory overhead.
- The extra forward pass can be **overlapped with Expert Parallelism (EP) communication**, which
  the paper reports bounds the additional wall-clock overhead to **roughly 20%**.

Activation policy (per the paper):

- An automatic detection mechanism triggers a **short rollback** and enables Anticipatory Routing
  **only when a loss spike is detected**.
- After operating in this mode for a period, the system **reverts to standard training**.
- The paper reports this averts loss spikes with negligible overall overhead and without
  compromising model performance.

## Why this could fit torchtitan

torchtitan has most of the machinery this needs:

- MoE layers with a routing/selection step, plus Expert Parallelism (the dispatch/combine
  all-to-all that the paper targets for overlap).
- Pipeline parallelism scheduling, which is where the extra single forward pass could be
  interleaved.
- A checkpoint/rollback story that would need to be aware of the trigger path and the cached
  routing state.

## Discussion

Implementing this would require some modification to the execution flow in `train_step`.
Specifically, when a loss spike triggers the short rollback and Anticipatory Routing is enabled,
there needs to be a **forward-only pass whose sole purpose is to compute and record routing
information** — a step shape that does not exist in the current training loop.

Would the maintainers be open to this technique being developed in torchtitan?
I'm happy to work on it if there's interest, and to align on the design before writing any code.

## References

- DeepSeek-V4 technical [report](https://arxiv.org/pdf/2606.19348) — Anticipatory Routing section.

**Comment by jojoinfra** [comment]
Thanks to @chango9543 for the RFC and @Dyslex7c for getting the mechanism up.

I've been developing this feature as well, and wanted to share where I think things stand.

As I read it, #4495 implements the cache/replay mechanism with a fixed one-step delay. The RFC also describes the dynamic part — detecting a loss spike, rolling back, running in that mode for a period, then reverting — which I don't see in the PR yet. That part is what keeps the cost down: the paper reports roughly 20% overhead while the mode is active (after overlapping the extra forward with EP communication), and says the overall cost is negligible only because the mode is armed on demand rather than always on.

The pieces that look outstanding to me:

- configurable delay, rather than a fixed one step
- loss-spike detection
- rollback to an earlier checkpoint
- the warmup / active / drain schedule around it

I wanted to ask whether these remaining pieces are already planned or in progress, so I don't duplicate anyone's effort.

**Comment by Dyslex7c** [comment]
Hey @jojoinfra, thanks for checking in and coordinating!

My PR (#4495) is limited to the **core mechanism** (the MoE cache/replay primitive and trainer execution for `dt=1`).

I am not currently working on the remaining pieces you mentioned (loss-spike detection, checkpoint rollback, the warmup/active/drain schedule, or configurable delay). They were left as follow-up work, so you can take them up or I'm also open to collaboration.

**Comment by jojoinfra** [comment]
Thanks @Dyslex7c for clarifying, and thanks for working on the core mechanism.

I'll pick up the remaining pieces and continue working toward completing the full Anticipatory Routing feature. Happy to coordinate as the work progresses.

**Comment by jojoinfra** [comment]
Following up on the discussion, I'd like to propose the intended behavior for Anticipatory Routing (AR) before discussing implementation details: enable AR after a loss spike triggers a checkpoint rollback, run it for a configurable number of optimizer steps, then return to normal training.

### Proposed flow

1. **Detect and roll back.** A configurable loss-spike detector would trigger rollback to a resumable checkpoint before the estimated spike onset. Model, optimizer, learning-rate scheduler, and dataloader state should be restored together to ensure a consistent starting point.

2. **Warm up the routing cache.** Prefetch future batches from a single data iterator and capture their per-layer expert IDs through no-grad forwards, without optimizer updates. Keep each batch together with its routing indices in a bounded queue.

3. **Train with delayed routing.** At each optimizer step, precompute routing for future data, then train on the oldest queued data. Replay only expert IDs; features and gating weights would still be computed using current parameters, with gradients flowing through the current model. The routing lag would ramp up initially and reach a configurable Δ optimizer steps in steady state.

4. **Return to normal training.** After the configured number of AR steps, stop prefetching and train the remaining queued batches with freshly computed routing. Once the queue is empty, resume the normal data flow. This preserves data order without discarding prefetched batches.

### Checkpointing and tradeoffs

For the initial design, I'd suggest deferring checkpoint saves during AR and draining, then resuming them in normal mode. Otherwise, the saved dataloader position could be ahead of completed training unless the pending batches and routing cache were also checkpointed. The tradeoff is a longer interval between recoverable checkpoints.

This approach keeps batches and routing indices aligned without requiring a second model copy or two independently advancing data streams. Additional routing forwards are needed only during warmup and active AR, rather than throughout training.

Would this be a reasonable starting point for the feature? I'd appreciate any feedback or alternative approaches.

**Comment by jojoinfra** [comment]
For the implementation, I'd like to propose a minimally invasive approach that builds on TorchTitan's existing extension mechanisms and reuses the core training and MoE execution paths.

### Extend the existing Trainer without copying the training loop

Following the [Trainer extension example](https://github.com/pytorch/torchtitan/blob/main/docs/extension.md#extending-trainerconfig), I'd introduce an `AnticipatoryTrainer` subclass under `experiments/anticipatory_routing`, with a corresponding `Trainer.Config` subclass for AR-specific settings.

This Trainer would inherit the existing training loop and use a dedicated schedule to manage AR's data flow and phase transitions. Since the existing `train_step()` accepts a microbatch iterator, the subclass can let the schedule supply the microbatches for the current optimizer step, then delegate execution to the base implementation:

```python
class AnticipatoryTrainer(Trainer):
    def train_step(
        self, data_iterator: Iterator[TrainingMicrobatch]
    ) -> None:
        with self.schedule.step(data_iterator) as microbatches:
            super().train_step(microbatches)
```

Before the base step runs, the schedule handles prefetching, selects the data to train on, and enters the appropriate routing mode. Afterward, it handles phase transitions and spike response. The supplied iterator contains exactly one optimizer step's microbatches and checks that they are all consumed.

This keeps the base `Trainer.train()` and `Trainer.train_step()` implementations unchanged, reusing their gradient-accumulation, optimizer-update, and logging logic.

An AR-specific Engine subclass would be selected through the existing `engine_cls` attribute. It would select each microbatch's routing-cache slot before delegating to `super().forward_backward_microbatch()`, while providing forward-only passes and rollback as separate operations.

### Extend expert selection through config overrides

On the model side, I'd use the documented [configurable override mechanism](https://github.com/pytorch/torchtitan/blob/main/torchtitan/overrides/README.md), activated through an opt-in `override.imports` entry.

An `@override` factory targeting `TokenChoiceTopKRouter.Config` would replace each matched router config with one that builds an `AnticipatoryRoutingMixin` combined with the original model-specific router. Using `derive()` would preserve the existing configuration fields rather than manually copying them.

The mixin would intercept expert selection through `_select_experts()`:

- **OFF:** delegate to `super()._select_experts()`.
- **CAPTURE:** delegate to the same method and cache the resulting expert IDs.
- **REPLAY:** supply the cached expert IDs instead of making a new selection.

Delegating through `super()` preserves the underlying router's selection logic during capture, including DeepSeek-V3's group-limited routing, without duplicating it.

The original router forward would still compute current scores and gather gating weights using the selected IDs, retaining the gradient path through current parameters. Routing-map construction, expert dispatch, expert computation, and combine would also reuse the existing implementation. No AR-specific branches would be needed in the core Router or `MoE.forward()`.

The aim is to keep AR-specific scheduling and capture/replay within the experiment, without maintaining separate copies of the core training or MoE logic as upstream evolves.

Would this be a reasonable starting point for the implementation?

@tianyu-l  — I'd appreciate your thoughts on the overall proposal. Any feedback, concerns, or suggestions for alternative approaches would be very welcome.

**Comment by tianyu-l** [comment]
thanks @jojoinfra ! This week occupied with other things, will get back whenever I can
